lib/streamlit/elements/data_editor.py

```diff
@@ -0,0 +1,582 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import contextlib
+import json
+from dataclasses import dataclass
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+)
+
+import pandas as pd
+import pyarrow as pa
+from pandas.api.types import is_datetime64_any_dtype, is_float_dtype, is_integer_dtype
+from pandas.io.formats.style import Styler
+from typing_extensions import Final, Literal, TypeAlias, TypedDict
+
+from streamlit import type_util
+from streamlit.elements.arrow import marshall
+from streamlit.elements.form import current_form_id
+from streamlit.errors import StreamlitAPIException
+from streamlit.proto.Arrow_pb2 import Arrow as ArrowProto
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.runtime.state import (
+    WidgetArgs,
+    WidgetCallback,
+    WidgetKwargs,
+    register_widget,
+)
+from streamlit.type_util import DataFormat, DataFrameGenericAlias, Key, to_key
+
+if TYPE_CHECKING:
+    import numpy as np
+
+    from streamlit.delta_generator import DeltaGenerator
+
+_INDEX_IDENTIFIER: Final = "index"
+
+# All formats that support direct editing, meaning that these
+# formats will be returned with the same type when used with data_editor.
+EditableData = TypeVar(
+    "EditableData",
+    bound=Union[
+        DataFrameGenericAlias[Any],  # covers DataFrame and Series
+        Tuple[Any],
+        List[Any],
+        Set[Any],
```

Comment:
Should a comment be noted that the backend is still a set but the frontend still adds a row?

Reply:
Yep, I think this wouldn't belong here and it's more of a temporary problem. I'm planning to app a uniqueness column configuration so that the frontend validates the value for uniqueness in a column.